File: Bridge/Resources/Exception.js
    // @source Exception.js

    Bridge.define("Bridge.Exception", {
        constructor: function (message, innerException) {
            this.message = message ? message : null;
            this.innerException = innerException ? innerException : null;
            this.errorStack = new Error();
            this.data = new Bridge.Dictionary$2(Object, Object)();
        },

        getMessage: function () {
            return this.message;
        },

        getInnerException: function () {
            return this.innerException;
        },

        getStackTrace: function () {
            return this.errorStack.stack;
        },

        getData: function () {
            return this.data;
        },

        toString: function () {
            return this.getMessage();
        },

        statics: {
            create: function (error) {
                if (Bridge.is(error, Bridge.Exception)) {
                    return error;
                }

                if (error instanceof TypeError) {
                    return new Bridge.NullReferenceException(error.message, new Bridge.ErrorException(error));
                } else if (error instanceof RangeError) {
                    return new Bridge.ArgumentOutOfRangeException(null, error.message, new Bridge.ErrorException(error));
                } else if (error instanceof Error) {
                    return new Bridge.ErrorException(error);
                } else {
                    return new Bridge.Exception(error ? error.toString() : null);
                }
            }
        }
    });

    Bridge.define("Bridge.ErrorException", {
        inherits: [Bridge.Exception],

        constructor: function (error) {
            Bridge.Exception.prototype.$constructor.call(this, error.message);
            this.errorStack = error;
            this.error = error;
        },

        getError: function () {
            return this.error;
        }
    });

    Bridge.define("Bridge.ArgumentException", {
        inherits: [Bridge.Exception],

        constructor: function (message, paramName, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Value does not fall within the expected range.", innerException);
            this.paramName = paramName ? paramName : null;
        },

        getParamName: function () {
            return this.paramName;
        }
    });

    Bridge.define("Bridge.ArgumentNullException", {
        inherits: [Bridge.ArgumentException],

        constructor: function (paramName, message, innerException) {
            if (!message) {
                message = "Value cannot be null.";

                if (paramName) {
                    message += "\nParameter name: " + paramName;
                }
            }

            Bridge.ArgumentException.prototype.$constructor.call(this, message, paramName, innerException);
        }
    });

    Bridge.define("Bridge.ArgumentOutOfRangeException", {
        inherits: [Bridge.ArgumentException],

        constructor: function (paramName, message, innerException, actualValue) {
            if (!message) {
                message = "Value is out of range.";

                if (paramName) {
                    message += "\nParameter name: " + paramName;
                }
            }

            Bridge.ArgumentException.prototype.$constructor.call(this, message, paramName, innerException);

            this.actualValue = actualValue ? actualValue : null;
        },

        getActualValue: function () {
            return this.actualValue;
        }
    });

    Bridge.define("Bridge.CultureNotFoundException", {
        inherits: [Bridge.ArgumentException],

        constructor: function (paramName, invalidCultureName, message, innerException, invalidCultureId) {
            if (!message) {
                message = "Culture is not supported.";

                if (paramName) {
                    message += "\nParameter name: " + paramName;
                }

                if (invalidCultureName) {
                    message += "\n" + invalidCultureName + " is an invalid culture identifier.";
                }
            }

            Bridge.ArgumentException.prototype.$constructor.call(this, message, paramName, innerException);

            this.invalidCultureName = invalidCultureName ? invalidCultureName : null;
            this.invalidCultureId = invalidCultureId ? invalidCultureId : null;
        },

        getInvalidCultureName: function () {
            return this.invalidCultureName;
        },

        getInvalidCultureId: function () {
            return this.invalidCultureId;
        }
    });

    Bridge.define("Bridge.KeyNotFoundException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Key not found.", innerException);
        }
    });

    Bridge.define("Bridge.ArithmeticException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Overflow or underflow in the arithmetic operation.", innerException);
        }
    });

    Bridge.define("Bridge.DivideByZeroException", {
        inherits: [Bridge.ArithmeticException],

        constructor: function (message, innerException) {
            Bridge.ArithmeticException.prototype.$constructor.call(this, message || "Division by 0.", innerException);
        }
    });

    Bridge.define("Bridge.OverflowException", {
        inherits: [Bridge.ArithmeticException],

        constructor: function (message, innerException) {
            Bridge.ArithmeticException.prototype.$constructor.call(this, message || "Arithmetic operation resulted in an overflow.", innerException);
        }
    });

    Bridge.define("Bridge.FormatException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Invalid format.", innerException);
        }
    });

    Bridge.define("Bridge.InvalidCastException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "The cast is not valid.", innerException);
        }
    });

    Bridge.define("Bridge.InvalidOperationException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Operation is not valid due to the current state of the object.", innerException);
        }
    });

    Bridge.define("Bridge.NotImplementedException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "The method or operation is not implemented.", innerException);
        }
    });

    Bridge.define("Bridge.NotSupportedException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Specified method is not supported.", innerException);
        }
    });

    Bridge.define("Bridge.NullReferenceException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Object is null.", innerException);
        }
    });

    Bridge.define("Bridge.RankException", {
        inherits: [Bridge.Exception],

        constructor: function (message, innerException) {
            Bridge.Exception.prototype.$constructor.call(this, message || "Attempted to operate on an array with the incorrect number of dimensions.", innerException);
        }
    });

File: Sample/__bin/coresystem.lua/CoreSystem/Exception.lua
local System = System

System.define("System.Exception", {
    __ctor__ = function(this, message, innerException) 
        this.message = message
        this.innerException = innerException
    end,

    getMessage = function(this) 
        return this.message
    end,

    getInnerException = function(this) 
        return this.innerException
    end,

    getStackTrace = function(this) 
        return this.errorStack
    end,

    toString = function(this) 
        return this.message .. this.errorStack
    end,

    traceback = function(this, lv)
        this.errorStack = debug.traceback("", lv and lv + 3 or 3)
    end,
})

System.define("System.ArgumentException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, paramName, innerException) 
        System.Exception.__ctor__(this, message or "Value does not fall within the expected range.", innerException)
        this.paramName = paramName
    end,

    getParamName = function(this) 
        return this.paramName
    end
})

System.define("System.ArgumentNullException", {
    __inherits__ = { System.ArgumentException },

    __ctor__ = function(this, paramName, message, innerException) 
        if not message then
            message = "Value cannot be null."
            if paramName then 
                message = message .. "\nParameter name = " .. paramName
            end
        end
        System.ArgumentException.__ctor__(this, message, paramName, innerException)
    end
})

System.define("System.ArgumentOutOfRangeException", {
    __inherits__ = { System.ArgumentException },

    __ctor__ = function(this, paramName, message, innerException, actualValue) 
        if not message then
            message = "Value is out of range."
            if paramName then
                message = message .. "\nParameter name = " .. paramName
            end
        end
        System.ArgumentException.__ctor__(this, message, paramName, innerException)
        this.actualValue = actualValue
    end,

    getActualValue = function(this) 
        return this.actualValue
    end
})

System.define("System.CultureNotFoundException", {
    __inherits__ = { System.ArgumentException },

    __ctor__ = function(this, paramName, invalidCultureName, message, innerException, invalidCultureId) 
        if not message then 
            message = "Culture is not supported."
            if paramName then
                message = message .. "\nParameter name = " .. paramName
            end
            if invalidCultureName then
                message = message .. "\n" .. invalidCultureName .. " is an invalid culture identifier."
            end
        end
        System.ArgumentException.__ctor__(this, message, paramName, innerException)
        this.invalidCultureName = invalidCultureName
        this.invalidCultureId = invalidCultureId
    end,

    getInvalidCultureName = function(this)
        return this.invalidCultureName
    end,

    getInvalidCultureId = function(this) 
        return this.invalidCultureId
    end
})

System.define("System.KeyNotFoundException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Key not found.", innerException)
    end
})

System.define("System.ArithmeticException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Overflow or underflow in the arithmetic operation.", innerException)
    end
})

System.define("System.DivideByZeroException", {
    __inherits__ = { System.ArithmeticException },

    __ctor__ = function(this, message, innerException) 
        System.ArithmeticException.__ctor__(this, message or "Division by 0.", innerException)
    end
})

System.define("System.OverflowException", {
    __inherits__ = { System.ArithmeticException },

    __ctor__ = function(this, message, innerException) 
        System.ArithmeticException.__ctor__(this, message or "Arithmetic operation resulted in an overflow.", innerException)
    end
})

System.define("System.FormatException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Invalid format.", innerException)
    end
})

System.define("System.InvalidCastException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "The cast is not valid.", innerException)
    end
})

System.define("System.InvalidOperationException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Operation is not valid due to the current state of the object.", innerException)
    end
})

System.define("System.NotImplementedException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "The method or operation is not implemented.", innerException)
    end
})

System.define("System.NotSupportedException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Specified method is not supported.", innerException)
    end
})

System.define("System.NullReferenceException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Object is null.", innerException)
    end
})

System.define("System.RankException", {
    __inherits__ = { System.Exception },

    __ctor__ = function(this, message, innerException) 
        System.Exception.__ctor__(this, message or "Attempted to operate on an array with the incorrect number of dimensions.", innerException)
    end
})
